tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -18,85 +18,267 @@ limitations under the License.
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_x);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_workgroup_id_x);
+      TargetInfo nvptx_info(nvptx_intrinsic_info, default_nvptx_function_info);
+      TargetInfo amdgpu_info(amdgpu_intrinsic_info, default_amdgpu_function_info);
+      return MultipleTargetInfo(nvptx_info, amdgpu_info);
     }
-    case TargetIntrinsicID::kBlockIdy: {
-      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_y,
-              llvm::Intrinsic::amdgcn_workgroup_id_y};
+    case TargetFunctionID::kBlockIdy: {
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_y);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_workgroup_id_y);
+      TargetInfo nvptx_info(nvptx_intrinsic_info, default_nvptx_function_info);
+      TargetInfo amdgpu_info(amdgpu_intrinsic_info, default_amdgpu_function_info);
+      return MultipleTargetInfo(nvptx_info, amdgpu_info);
     }
-    case TargetIntrinsicID::kBlockIdz: {
-      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z,
-              llvm::Intrinsic::amdgcn_workgroup_id_z};
+    case TargetFunctionID::kBlockIdz: {
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_workgroup_id_z);
+      TargetInfo nvptx_info(nvptx_intrinsic_info, default_nvptx_function_info);
+      TargetInfo amdgpu_info(amdgpu_intrinsic_info, default_amdgpu_function_info);
+      return MultipleTargetInfo(nvptx_info, amdgpu_info);
     }
-    case TargetIntrinsicID::kBarrierId: {
-      return {llvm::Intrinsic::nvvm_barrier0,
-              llvm::Intrinsic::amdgcn_s_barrier};
+    case TargetFunctionID::kBarrierId: {
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_barrier0);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_s_barrier);
+      TargetInfo nvptx_info(nvptx_intrinsic_info, default_nvptx_function_info);
+      TargetInfo amdgpu_info(amdgpu_intrinsic_info, default_amdgpu_function_info);
+      return MultipleTargetInfo(nvptx_info, amdgpu_info);
     }
   }
 }
 }  // namespace
 
-llvm::CallInst* EmitCallToTargetIntrinsic(
-    TargetIntrinsicID intrinsic_id, absl::Span<llvm::Value* const> operands,
-    absl::Span<llvm::Type* const> overloaded_types, llvm::IRBuilder<>* b) {
+llvm::Value* EmitCallToTargetFunction(
+    TargetFunctionID function_id, absl::Span<llvm::Value* const> operands,
+    absl::Span<const PrimitiveType> input_types, PrimitiveType output_type,
+    absl::Span<const llvm::Attribute::AttrKind> attributes,
+    absl::Span<llvm::Type* const> overloaded_types,
+    llvm::IRBuilder<>* b) {
   llvm::Module* module = b->GetInsertBlock()->getModule();
-  struct TargetIntrinsics gpu_intrinsic_id = GetIntrinsic(intrinsic_id);
+  struct MultipleTargetInfo all_gpu_info = GetTargetInfo(function_id);
```

Comment:
@sumesh13 `MultipleTargetInfo` is actually a dual-target info which may possess limitation in the long run.

Would you consider passing `target_triple` to `GetTargetInfo` and return a `TargetInfo` directly? Then we can eliminate `MultipleTargetInfo` altogether.

Reply:
Sure..thats worth a try. Did you want them interspersed with the switch-case - that makes it too many branches  - or just at the end after both nvptx/amdgpu structures have been populated. The later will still let us avoid this structure.

Reply:
Whichever way you feel like. My intention is to remove this extra class which shouldn’t really be needed by clients.